src/chunk.c

```diff
@@ -41,12 +47,242 @@ chunk_create_from_tuple(HeapTuple tuple, int16 num_constraints)
 	return chunk;
 }
 
+/*
+ * Align a chunk's hypercube in 'aligned' dimensions.
+ *
+ * After a chunk collision scan, this function is called for each chunk in the
+ * chunk scan context. Chunks in the scan context may have only a partial set of
+ * slices if they only overlap in some, but not all, dimensions (see
+ * illustrations below). Still, partial chunks may still be of interest for
+ * alignment in a particular dimension. Thus, if a chunk has an overlapping
+ * slice in an aligned dimension, we cut the context's hypercube to not overlap
+ * with that slice.
+ */
+static bool
+do_dimension_alignment(ChunkScanCtx *scanctx, Chunk *chunk)
+{
+	Hypercube  *cube = scanctx->data;
+	Hyperspace *space = scanctx->space;
+	int			i;
+
+	for (i = 0; i < space->num_dimensions; i++)
+	{
+		Dimension  *dim = &space->dimensions[i];
+		DimensionSlice *chunk_slice,
+				   *cube_slice;
+		int64		coord = scanctx->point->coordinates[i];
+
+		if (!dim->fd.aligned)
+			continue;
+
+		/*
+		 * The chunk might not have a slice for each dimension, so fetch slice
+		 * by dimension ID rather than index
+		 */
+		chunk_slice = hypercube_get_slice_by_dimension_id(chunk->cube, dim->fd.id);
+
+		if (NULL == chunk_slice)
+			continue;
+
+		cube_slice = cube->slices[i];
+
+		/*
+		 * Only cut-to-align if the slices collide and are not identical
+		 * (i.e., if we are reusing an existing slice we should not cut it)
+		 */
+		if (!dimension_slices_equal(cube_slice, chunk_slice) &&
+			dimension_slices_collide(cube_slice, chunk_slice))
+			dimension_slice_cut(cube_slice, chunk_slice, coord);
+	}
+
+	return true;
+}
+
+/*
+ * Resolve chunk collisions.
+ *
+ * After a chunk collision scan, this function is called for each chunk in the
+ * chunk scan context. We only care about chunks that have a full set of
+ * slices/constraints that overlap with our tentative hypercube, i.e., they
+ * fully collide. We resolve those collisions by cutting the hypercube.
+ */
+static bool
+do_collision_resolution(ChunkScanCtx *scanctx, Chunk *chunk)
+{
+	Hypercube  *cube = scanctx->data;
+	Hyperspace *space = scanctx->space;
+	int			i;
+
+	if (chunk->cube->num_slices != space->num_dimensions ||
+		!hypercubes_collide(cube, chunk->cube))
+		return true;
+
+	for (i = 0; i < space->num_dimensions; i++)
+	{
+		DimensionSlice *cube_slice = cube->slices[i];
+		DimensionSlice *chunk_slice = chunk->cube->slices[i];
+		int64		coord = scanctx->point->coordinates[i];
+
+		/*
+		 * Only cut if we aren't reusing an existing slice and there is a
+		 * collision
+		 */
+		if (!dimension_slices_equal(cube_slice, chunk_slice) &&
```

Comment:
Shouldn't we be allowed to cut equal slices for non-aligned dimensions? i.e. Shouldn't this check not apply when this dimension is not-aligned?

Reply:
Actually I think we can skip this entirely if dimension is aligned, no? (handled in do_dimension_alignment?)

Reply:
I don't see a way to reasonably cut one out of two equal slices so that they no longer overlap (unless you make the range of one of them 0). Look at `dimension_slice_cut()` to check for the conditions where a cut actually happens. You will find that in the case of equal slice it won't cut anyway so this is just an extra check.

Reply:
yeah nevermind.